src/device/linux-arm-rpi2/gpioctl_linux-arm-rpi2.cpp

```diff
@@ -0,0 +1,295 @@
+public:
+  explicit GpioControlInst(JObject& jgpioctl);
+  virtual int Initialize(void);
+  virtual void Release(void);
+  virtual int PinMode(uint32_t portpin);
+  virtual int WritePin(uint32_t portpin, uint8_t data);
+  virtual int ReadPin(uint32_t portpin, uint8_t* pdata);
+
+protected:
+  int open_device(const char* devicefile, bool wmode);
+  int enable_port(uint8_t gpionum, bool enable);
+};
+
+
+//-----------------------------------------------------------------------------
+// portpin mode is 32bit integer where bits are used as below
+//
+//   31 ~ 24 : not used
+//   23 ~ 16 : mode
+//   15 ~ 08 : not used
+//   07 ~ 00 : pin
+//
+
+#define GPIO_RPI2_MODE_ENABLE     0x00020000
+#define GPIO_RPI2_MODE_OUTPUT     0x00010000
+
+// documents:
+//   https://www.raspberrypi.org/documentation/usage/gpio/
+//   https://www.kernel.org/doc/Documentation/gpio/sysfs.txt
+
+static const char* gpio_export_1 = "/sys/class/gpio/export";
+static const char* gpio_export_0 = "/sys/class/gpio/unexport";
+static const char* gpio_dirfmt   = "/sys/class/gpio/gpio%d/direction";
+static const char* gpio_valfmt   = "/sys/class/gpio/gpio%d/value";
+
+
+#define GPIO_RPI2_MAX_PATH        40
+
+static char gpio_file_buff[GPIO_RPI2_MAX_PATH];
+
+
+//-----------------------------------------------------------------------------
+// table to convert pin number to gpio number
+//
+static const uint8_t rpi2_pintogpio[41] = {
+  0,
+  // 1 ~ 10
+  0, 0,
+  2, 0,
+  3, 0,
+  4, 14,
+  0, 15,
+  // 11 ~ 20
+  17, 18,
+  27, 0,
+  22, 23,
+  0, 24,
+  10, 0,
+  // 21 ~ 30
+  9, 25,
+  11, 8,
+  0, 7,
+  0, 0,
+  5, 0,
+  // 31 ~ 40
+  6, 12,
+  13, 0,
+  19, 16,
+  26, 20,
+  0, 21
+};
+
+static uint8_t rpi2_get_gpio(uint8_t pin) {
+  if (pin > 40)
+    return 0;
+  return rpi2_pintogpio[pin];
+}
+
+
+//-----------------------------------------------------------------------------
+
+GpioControl* GpioControl::Create(JObject& jgpioctl)
+{
+  return new GpioControlInst(jgpioctl);
+}
+
+
+GpioControlInst::GpioControlInst(JObject& jgpioctl)
+    : GpioControl(jgpioctl) {
+}
+
+
+int GpioControlInst::Initialize(void) {
+  if (_fd > 0 )
+    return IOTJS_GPIO_INUSE;
+
+  // there's actually no initializing needed for RPi2
+  // just check write access rights for gpio device file
+  int fd = open_device(gpio_export_1, true);
+  if (fd < 0) {
+    _fd = -1;
+    DLOG("RPi2 GPIO failed to initialize, check account gpio access rights");
+    return IOTJS_GPIO_INITFAIL;
+  }
+  close(fd);
+
+  _fd = 1;
+  return _fd;
+}
+
+
+void GpioControlInst::Release(void) {
+  _fd = 0;
+}
+
+
+#define GPIO_OPEN_RETRY           30
+#define GPIO_RETRY_WAIT           1000      // in microseconds
+
+int GpioControlInst::open_device(const char* devicefile, bool wmode) {
+  int fd = -1;
+  int retry = GPIO_OPEN_RETRY;
+  int sleepvalue = GPIO_RETRY_WAIT;
+  int flags = wmode ? O_WRONLY : O_RDONLY;
+
+  // creating virtual file for device control takes some time
+  while (fd < 0 && retry > 0) {
```

Comment:
I think 0.5 sec for RPi2 is ok, for other boards there will be signal specification that we can specify.